chapter07/src/main/java/com/seaofnodes/simple/Parser.java

```diff
@@ -134,14 +135,60 @@ private Node parseBlock() {
      * @return a {@link Node} or {@code null}
      */
     private Node parseStatement() {
-        if (matchx("return")  ) return parseReturn();
-        else if (matchx("int")) return parseDecl();
-        else if (match ("{"  )) return require(parseBlock(),"}");
-        else if (matchx("if" )) return parseIf();
+        if (matchx("return")    ) return parseReturn();
+        else if (matchx("int"  )) return parseDecl();
+        else if (match ("{"    )) return require(parseBlock(),"}");
+        else if (matchx("if"   )) return parseIf();
+        else if (matchx("while")) return parseWhile();
         else if (matchx("#showGraph")) return require(showGraph(),";");
         else return parseExpressionStatement();
     }
 
+    /**
+     * Parses a while statement
+     *
+     * <pre>
+     *     while ( expression ) statement
+     * </pre>
+     * @return a {@link Node}, never {@code null}
+     */
+    private Node parseWhile() {
+        require("(");
+
+        RegionNode region = (RegionNode) ctrl(new RegionNode(null, ctrl(), null).peephole());
+        ctrl(region);
+
+        ScopeNode outerScope = _scope;
+        ScopeNode loopScope = outerScope.makeLoopScope();
+        _allScopes.push(loopScope);
+        _scope = loopScope;
+
+        // Parse predicate
+        var pred = require(parseExpression(), ")");
+        // IfNode takes current control and predicate
+        IfNode ifNode = (IfNode)new IfNode(ctrl(), pred).<IfNode>keep().peephole();
+        // Setup projection nodes
+        Node ifT = new ProjNode(ifNode, 0, "True" ).peephole();
+        ifNode.unkeep();
+        Node ifF = new ProjNode(ifNode, 1, "False").peephole();
+
+        // Parse the true side, which corresponds to loop body
+        ctrl(ifT);              // set ctrl token to ifTrue projection
+        parseStatement();       // Parse loop body
+
+        if( outerScope.nIns() != _scope.nIns() )
+            throw error("Cannot define a new name in a while loop");
```

Comment:
well, you ought to be able to define one, with scope just in the loop body.
```java
while( x < lim ) {
  int tmp0 = compute();
  int tmp1 = more();
  x = new_x(tmp0,tmp1);
}         
```

Reply:
Yes, this is to catch
```c
while(...) int a = 1;
return a;
```